# Análise de dados estruturados

In [1]:
import pandas as pd

![panda](https://media.giphy.com/media/HDR31jsQUPqQo/giphy.gif)

## Criar um dataframe a partir de um dicionário

In [2]:
dict = {"country": ["Brazil", "Russia", "India", "China", "South Africa"],
       "capital": ["Brasilia", "Moscow", "New Dehli", "Beijing", "Pretoria"],
       "area": [8.516, 17.10, 3.286, 9.597, 1.221],
       "population": [200.4, 143.5, 1252, 1357, 52.98] }

In [3]:
# Transformar o dicionário em um dataframe
brics = pd.DataFrame(dict)

In [4]:
# Ver os primeiros registros desse dataframe
brics.head()

area    capital       country  population
0   8.516   Brasilia        Brazil      200.40
1  17.100     Moscow        Russia      143.50
2   3.286  New Dehli         India     1252.00
3   9.597    Beijing         China     1357.00
4   1.221   Pretoria  South Africa       52.98

## Importar um csv com o pandas

Vamos utilizar os dados que o Kaggle lançou no ano de 2017 sobre Cientistas de Dados e Data Science. São 5 datasets diferentes:

 - **schema.csv**: a CSV file with survey schema. This schema includes the questions that correspond to each column name in both the multipleChoiceResponses.csv and freeformResponses.csv.
 - **multipleChoiceResponses.csv**: Respondents' answers to multiple choice and ranking questions. These are non-randomized and thus a single row does correspond to all of a single user's answers. 
 -**freeformResponses.csv:** Respondents' freeform answers to Kaggle's survey questions. These responses are randomized within a column, so that reading across a single row does not give a single user's answers.
 - **conversionRates.csv**: Currency conversion rates (to USD) as accessed from the R package "quantmod" on September 14, 2017
 - **RespondentTypeREADME.txt**: This is a schema for decoding the responses in the "Asked" column of the schema.csv file.

In [4]:
# Carregue o dataset multipleChoiceResponses com o pandas 
multiple_choice = pd.read_csv('kaggle-survey-2017/multipleChoiceResponses.csv')

In [5]:
# Veja as primeiras linhas do dataset
multiple_choice.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit              ...                \
0             DBA/Database Engineer     Fine              ...                 
1                               NaN      NaN              ...                 
2                               NaN      NaN              ...                 
3  Operations Research Practitioner   Poorly              ...                 
4                Computer Scientist     Fine              ...                 

  JobFactorExperienceLevel JobFactorDepartment  JobFactorTitle  \
0                      NaN                 NaN             NaN   
1                      NaN                 NaN             NaN   
2           Very Important      Very Important  Very Important   
3                      NaN                 NaN             NaN   
4                      NaN                 NaN             NaN   

  JobFactorCompanyFunding JobFactorImpact JobFactorRemote JobFactorIndustry  \
0                     NaN             NaN             NaN               NaN   
1                     NaN             NaN             NaN               NaN   
2          Very Important  Very Important  Very Important    Very Important   
3                     NaN             NaN             NaN               NaN   
4                     NaN             NaN             NaN               NaN   

  JobFactorLeaderReputation JobFactorDiversity JobFactorPublishingOpportunity  
0                       NaN                NaN                            NaN  
1        Somewhat important                NaN                            NaN  
2            Very Important     Very Important                 Very Important  
3                       NaN                NaN                            NaN  
4                       NaN                NaN                            NaN  

[5 rows x 228 columns]

In [7]:
#Veja a quantidade de linhas e de colunas do dataset
multiple_choice.shape

(16716, 228)

Existem 228 colunas!!!
![panda](https://media.giphy.com/media/14aUO0Mf7dWDXW/giphy.gif)

Vamos ver do que se tratam essas colunas. Como são MUITAS colunas precisamos alterar a configuração padrão do pandas para visualização de linhas e colunas

In [8]:
pd.set_option('max_rows', 200)
pd.set_option('max_columns', 1000)

In [9]:
multiple_choice.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit  \
0             DBA/Database Engineer     Fine   
1                               NaN      NaN   
2                               NaN      NaN   
3  Operations Research Practitioner   Poorly   
4                Computer Scientist     Fine   

                                 CurrentEmployerType MLToolNextYearSelect  \
0  Employed by a company that doesn't perform adv...             SAS Base   
1                                                NaN               Python   
2                                                NaN  Amazon Web services   
3                                      Self-employed           TensorFlow   
4  Employed by a company that doesn't perform adv...           TensorFlow   

  MLMethodNextYearSelect LanguageRecommendationSelect  \
0         Random Forests                           F#   
1         Random Forests                       Python   
2          Deep learning                            R   
3            Neural Nets                       Python   
4            Text Mining                       Python   

                                PublicDatasetsSelect  \
0  Dataset aggregator/platform (i.e. Socrata/Kagg...   
1  Dataset aggregator/platform (i.e. Socrata/Kagg...   
2  Dataset aggregator/platform (i.e. Socrata/Kagg...   
3          I collect my own data (e.g. web-scraping)   
4                                             GitHub   

                              LearningPlatformSelect  \
0  College/University,Conferences,Podcasts,Trade ...   
1                                             Kaggle   
2  Arxiv,College/University,Kaggle,Online courses...   
3  Blogs,College/University,Conferences,Friends n...   
4                  Arxiv,Conferences,Kaggle,Textbook   

  LearningPlatformUsefulnessArxiv LearningPlatformUsefulnessBlogs  \
0                             NaN                             NaN   
1                             NaN                             NaN   
2                     Very useful                             NaN   
3                             NaN                     Very useful   
4                     Very useful                             NaN   

  LearningPlatformUsefulnessCollege LearningPlatformUsefulnessCompany  \
0                               NaN                               NaN   
1                               NaN                               NaN   
2                   Somewhat useful                               NaN   
3                       Very useful                               NaN   
4                               NaN                               NaN   

  LearningPlatformUsefulnessConferences LearningPlatformUsefulnessFriends  \
0                           Very useful                               NaN   
1                                   NaN                               Na

Podemos ver só o nome das colunas também utilizando o `columns`. Para ficar mais fácil de visualizar, ao invés de retornar o array, podemos transformar esse dado em uma Series.

In [10]:
# Use o columns no dataframe e coloque-o em uma Series para facilitar a visualização
pd.Series(multiple_choice.columns)

0                                     GenderSelect
1                                          Country
2                                              Age
3                                 EmploymentStatus
4                                    StudentStatus
5                              LearningDataScience
6                                       CodeWriter
7                                   CareerSwitcher
8                            CurrentJobTitleSelect
9                                         TitleFit
10                             CurrentEmployerType
11                            MLToolNextYearSelect
12                          MLMethodNextYearSelect
13                    LanguageRecommendationSelect
14                            PublicDatasetsSelect
15                          LearningPlatformSelect
16                 LearningPlatformUsefulnessArxiv
17                 LearningPlatformUsefulnessBlogs
18               LearningPlatformUsefulnessCollege
19               LearningPlatfo

Podemos ver mais detalhes do dataset com o `info()`

In [11]:
multiple_choice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16716 entries, 0 to 16715
Columns: 228 entries, GenderSelect to JobFactorPublishingOpportunity
dtypes: float64(13), object(215)
memory usage: 29.1+ MB


Podemos dar uma olhada nos tipos de campos que vem em cada uma das colunas númericas com um único comando

In [12]:
multiple_choice.describe()

Age  LearningCategorySelftTaught  \
count  16385.000000                 13109.000000   
mean      32.372841                    33.366771   
std       10.473487                    25.787181   
min        0.000000                     0.000000   
25%       25.000000                    15.000000   
50%       30.000000                    30.000000   
75%       37.000000                    50.000000   
max      100.000000                   100.000000   

       LearningCategoryOnlineCourses  LearningCategoryWork  \
count                   13126.000000          13111.000000   
mean                       27.375514             15.217593   
std                        26.860840             18.996778   
min                         0.000000              0.000000   
25%                         5.000000              0.000000   
50%                        20.000000             10.000000   
75%                        40.000000             25.000000   
max                       100.000000            100.000000   

       LearningCategoryUniversity  LearningCategoryKaggle  \
count                13122.000000            13126.000000   
mean                    16.988607                5.531434   
std                     23.676917               11.072680   
min                      0.000000                0.000000   
25%                      0.000000                0.000000   
50%                      5.000000                0.000000   
75%                     30.000000               10.000000   
max                    100.000000              100.000000   

       LearningCategoryOther  TimeGatheringData  TimeModelBuilding  \
count           13094.000000        7530.000000        7528.000000   
mean                1.795940          36.144754          21.268066   
std                 9.357886          21.649591          16.165958   
min                 0.000000           0.000000           0.000000   
25%                 0.000000          20.000000          10.000000   
50%                 0.000000          35.000000          20.000000   
75%                 0.000000          50.000000          30.000000   
max               100.000000         100.000000         100.000000   

       TimeProduction  TimeVisualizing  TimeFindingInsights  TimeOtherSelect  
count     7517.000000      7529.000000          7523.000000      7513.000000  
mean        10.806372        13.869372            13.094776         2.396247  
std         12.257932        11.722945            12.974846        12.157137  
min          0.000000         0.000000             0.000000         0.000000  
25%          0.000000         5.000000             5.000000         0.000000  
50%         10.000000        10.000000            10.000000         0.000000  
75%         15.000000        20.000000            20.000000         0.000000  
max        100.000000       100.000000           303.000000       100.000000

E se eu quiser ver a quantidade de nulos no dataset todo?

In [13]:
multiple_choice.isnull().sum()

GenderSelect                                      95
Country                                          121
Age                                              331
EmploymentStatus                                   0
StudentStatus                                  15436
LearningDataScience                            15432
CodeWriter                                      3530
CareerSwitcher                                 13704
CurrentJobTitleSelect                           4886
TitleFit                                        5212
CurrentEmployerType                             5115
MLToolNextYearSelect                            5718
MLMethodNextYearSelect                          5883
LanguageRecommendationSelect                    5718
PublicDatasetsSelect                            5920
LearningPlatformSelect                          5445
LearningPlatformUsefulnessArxiv                14325
LearningPlatformUsefulnessBlogs                11951
LearningPlatformUsefulnessCollege             

E se eu quiser fazer a porcentagem de nulos?

In [14]:
multiple_choice.isnull().sum() / len(multiple_choice)

GenderSelect                                   0.005683
Country                                        0.007239
Age                                            0.019801
EmploymentStatus                               0.000000
StudentStatus                                  0.923427
LearningDataScience                            0.923187
CodeWriter                                     0.211175
CareerSwitcher                                 0.819813
CurrentJobTitleSelect                          0.292295
TitleFit                                       0.311797
CurrentEmployerType                            0.305994
MLToolNextYearSelect                           0.342067
MLMethodNextYearSelect                         0.351938
LanguageRecommendationSelect                   0.342067
PublicDatasetsSelect                           0.354152
LearningPlatformSelect                         0.325736
LearningPlatformUsefulnessArxiv                0.856963
LearningPlatformUsefulnessBlogs                0

Nossa quanto nulo!
![sad_panda](https://media.giphy.com/media/3e18NPUVzoxzO/giphy.gif)

O que eu devo fazer se eu quiser ver apenas coluna `JobFactorSalary`?

In [15]:
multiple_choice['JobFactorSalary']

0                       NaN
1                       NaN
2            Very Important
3                       NaN
4                       NaN
5                       NaN
6                       NaN
7            Very Important
8                       NaN
9                       NaN
10       Somewhat important
11                      NaN
12       Somewhat important
13                      NaN
14                      NaN
15                      NaN
16                      NaN
17                      NaN
18       Somewhat important
19            Not important
20       Somewhat important
21                      NaN
22                      NaN
23                      NaN
24                      NaN
25                      NaN
26                      NaN
27                      NaN
28                      NaN
29           Very Important
30                      NaN
31                      NaN
32                      NaN
33                      NaN
34                      NaN
35           Very Im

Vamos fazer algumas operações com o pandas para contar o número de nulos que existem nessa coluna

In [16]:
multiple_choice['JobFactorSalary'].isnull().sum()

13231

Como eu faço se eu só quiser ver os 10 primeiros registros?

In [17]:
multiple_choice['JobFactorSalary'][:10]

0               NaN
1               NaN
2    Very Important
3               NaN
4               NaN
5               NaN
6               NaN
7    Very Important
8               NaN
9               NaN
Name: JobFactorSalary, dtype: object

E se eu quiser ver 2 colunas ao mesmo tempo? (E apenas essas 2 colunas)

In [18]:
multiple_choice[['JobFactorSalary', 'JobFactorLearning']][:10]

JobFactorSalary JobFactorLearning
0             NaN               NaN
1             NaN               NaN
2  Very Important    Very Important
3             NaN               NaN
4             NaN               NaN
5             NaN               NaN
6             NaN               NaN
7  Very Important    Very Important
8             NaN               NaN
9             NaN               NaN

O quanto que as pessoas dessa pesquisa estão satisfeitas com o trabalhos? Conseguimos saber isso usando só o pandas?

In [19]:
multiple_choice['JobSatisfaction'].value_counts()

7                          1448
8                          1427
6                           765
9                           677
5                           627
10 - Highly Satisfied       589
3                           358
4                           354
1 - Highly Dissatisfied     167
I prefer not to share       148
2                           117
Name: JobSatisfaction, dtype: int64

Percebemos com esse comando que as pessoas até que estão bastante satisfeitas.

Agora vamos olhar só as pessoas que estão Super Satisfeitas (Highly Satisfied) com o seu trabalho. Como que eu posso fazer isso?

In [20]:
# Filtre só quem está com o JobSatisfaction de 10. Guarde isso em uma variável pq é bastante dado
highly_satisfied = multiple_choice[multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied']

In [21]:
# veja o tamanho do dataset. Ele bateu com a quantidade de pessoas que estão altamente satisfeitas?
highly_satisfied.shape

(589, 228)

In [22]:
# Veja os primeiros 3 registros (todas as colunas) das pessoas altamentes satisfeitas
highly_satisfied[:3]

GenderSelect        Country   Age  \
3          Male  United States  56.0   
60         Male         Canada  34.0   
77         Male         Israel  29.0   

                                     EmploymentStatus StudentStatus  \
3   Independent contractor, freelancer, or self-em...           NaN   
60  Independent contractor, freelancer, or self-em...           NaN   
77  Independent contractor, freelancer, or self-em...           NaN   

   LearningDataScience CodeWriter CareerSwitcher  \
3                  NaN        Yes            NaN   
60                 NaN        Yes            NaN   
77                 NaN        Yes            NaN   

               CurrentJobTitleSelect   TitleFit CurrentEmployerType  \
3   Operations Research Practitioner     Poorly       Self-employed   
60                             Other       Fine       Self-employed   
77                    Data Scientist  Perfectly       Self-employed   

   MLToolNextYearSelect MLMethodNextYearSelect LanguageRecommendationSelect  \
3            TensorFlow            Neural Nets                       Python   
60                  NaN                    NaN                       Python   
77                  NaN          Deep learning                       Python   

                                 PublicDatasetsSelect  \
3           I collect my own data (e.g. web-scraping)   
60                                                NaN   
77  Dataset aggregator/platform (i.e. Socrata/Kagg...   

                               LearningPlatformSelect  \
3   Blogs,College/University,Conferences,Friends n...   
60  Arxiv,Blogs,Kaggle,Online courses,Personal Pro...   
77  Arxiv,Kaggle,Official documentation,Online cou...   

   LearningPlatformUsefulnessArxiv LearningPlatformUsefulnessBlogs  \
3                              NaN                     Very useful   
60                             NaN                             NaN   
77                 Somewhat useful                             NaN   

   LearningPlatformUsefulnessCollege LearningPlatformUsefulnessCompany  \
3                        Very useful                               NaN   
60                               NaN                               NaN   
77                               NaN                               NaN   

   LearningPlatformUsefulnessConferences LearningPlatformUsefulnessFriends  \
3                            Very useful                       Very useful   
60                                   NaN                               NaN   
77                                   NaN                               NaN   

   LearningPlatformUsefulnessKaggle LearningPlatformUsefulnessNewsletters  \
3                               NaN                                   NaN   
60                              NaN                                   NaN   
77                      Very useful                                   NaN   

   LearningPlatformUsefulnessCommunities  \
3                                    NaN   
60                                   NaN   
77                                   NaN   

   LearningPlatformUsefulnessDocumentation LearningPlatformUsefulnessCourses  \
3                              Very useful                       Very useful   
60                                     NaN                               NaN   
77                             Very useful                       Very useful   

   LearningPlatformUsefulnessProjects LearningPlatformUsefulnessPodcasts  \
3                         Very useful                                NaN   
60                                NaN                                NaN   
77                                NaN                                NaN   

   LearningPlatformUsefulnessSO LearningPlatformUsefulnessTextbook  \
3                           NaN                                NaN   
60                          NaN                                NaN   
77                  Very useful                                NaN   

   L

E se eu quiser ver as pessoas altamente satisfeitas e que trabalham com python?

In [23]:
highly_satisfied = multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied'
pythonist = multiple_choice['LanguageRecommendationSelect'] == 'Python'
highly_satisfied_and_pythonist = multiple_choice[highly_satisfied & pythonist]

In [24]:
highly_satisfied_and_pythonist.shape

(319, 228)

E se tentassemos com a idade? Ver só que está abaixo de 30 anos

In [25]:
highly_satisfied = multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied'
age = multiple_choice['Age'] < 30.0
highly_satisfied_and_age = multiple_choice[highly_satisfied & age]

In [26]:
highly_satisfied_and_age.shape

(171, 228)

Quais são as linguagens que a galera altamente satisfeita recomenda?

In [27]:
multiple_choice[highly_satisfied]['LanguageRecommendationSelect'].value_counts()

Python      319
R           151
SQL          28
C/C++/C#     20
Matlab       11
Other         9
Scala         9
Java          7
Julia         5
SAS           4
Stata         3
Haskell       2
Name: LanguageRecommendationSelect, dtype: int64

In [28]:
highly_satisfied_languages = multiple_choice[highly_satisfied]['LanguageRecommendationSelect'].value_counts()
language_counts = multiple_choice['JobSatisfaction'][highly_satisfied].notnull().sum()

In [29]:
language_counts

589

In [30]:
(highly_satisfied_languages / language_counts) * 100

Python      54.159593
R           25.636672
SQL          4.753820
C/C++/C#     3.395586
Matlab       1.867572
Other        1.528014
Scala        1.528014
Java         1.188455
Julia        0.848896
SAS          0.679117
Stata        0.509338
Haskell      0.339559
Name: LanguageRecommendationSelect, dtype: float64

E se eu quiser ordenar esses valores? Do menor para o maior?

In [31]:
pd.Series((highly_satisfied_languages / language_counts) * 100).sort_values()

Haskell      0.339559
Stata        0.509338
SAS          0.679117
Julia        0.848896
Java         1.188455
Other        1.528014
Scala        1.528014
Matlab       1.867572
C/C++/C#     3.395586
SQL          4.753820
R           25.636672
Python      54.159593
Name: LanguageRecommendationSelect, dtype: float64

### Desafio 1

Qual o país que tem a maior quantidade de dados onde as pessoas preencheram a coluna que tem o menor número dos dados?

Dica: Você precisará ordenar os campos pela quantidade de nulos (ou não nulos) e depois ver o país dessa galera.

![arrested_panda](https://media.giphy.com/media/N6funLtVsHW0g/giphy.gif)

In [32]:
multiple_choice.isnull().sum().sort_values(ascending=False)

WorkToolsFrequencyAngoss                       16694
WorkToolsFrequencySalfrod                      16684
WorkToolsFrequencyKNIMECommercial              16680
WorkMethodsFrequencySelect2                    16677
WorkToolsFrequencyStatistica                   16674
WorkToolsFrequencyDataRobot                    16653
WorkToolsFrequencyRapidMinerCommercial         16642
WorkFrequencySelect3                           16635
WorkToolsFrequencySAPBusinessObjects           16625
WorkMethodsFrequencySelect3                    16623
JobSkillImportanceOtherSelect3                 16603
WorkToolsFrequencySASJMP                       16601
WorkToolsFrequencyOrange                       16594
WorkToolsFrequencySelect2                      16581
WorkToolsFrequencyTIBCO                        16577
WorkToolsFrequencyFlume                        16575
WorkToolsFrequencyMinitab                      16572
WorkToolsFrequencyStan                         16564
WorkToolsFrequencyIBMCognos                   

In [33]:
multiple_choice[multiple_choice['WorkToolsFrequencyAngoss'].notnull()]['Country'].value_counts()

United States     8
India             3
Mexico            2
Italy             2
Canada            2
Singapore         1
Australia         1
Egypt             1
Other             1
United Kingdom    1
Name: Country, dtype: int64

## Selecionando por index

E se eu quiser pegar os valores de uma linha específica do dataframe?

In [34]:
multiple_choice.iloc[10,]

GenderSelect                                                                              Female
Country                                                                                   Russia
Age                                                                                           20
EmploymentStatus                                          Not employed, and not looking for work
StudentStatus                                                                                Yes
LearningDataScience                            Yes, I'm focused on learning mostly data scien...
CodeWriter                                                                                   NaN
CareerSwitcher                                                                               NaN
CurrentJobTitleSelect                                                                        NaN
TitleFit                                                                                     NaN
CurrentEmployerType           

Também posso ver só o valor de uma coluna, sem escrever o nome, somente pela sua posição

In [35]:
multiple_choice.iloc[:,0]

0        Non-binary, genderqueer, or gender non-conforming
1                                                   Female
2                                                     Male
3                                                     Male
4                                                     Male
5                                                     Male
6                                                     Male
7                                                   Female
8                                                   Female
9                                                     Male
10                                                  Female
11                                                    Male
12                                                    Male
13                                                    Male
14                                                    Male
15                                                    Male
16                                                    Ma

Mais detalhes sobre `loc`, `iloc` e `ix` podem ser vistas nesse [link](https://www.shanelynn.ie/select-pandas-dataframe-rows-and-columns-using-iloc-loc-and-ix/)

### Desafio 2

Será que temos uma quantidade gigante de nulos nesse dataset porque as pessoas não preencheram essas perguntas por multipla escolha, por que responderam no modo livre? 

Para validar essa hipótese teremos que carregar o outro dataset, que contém as perguntas em forma livre. e juntar (Pelo menos uma das variáveis) dos dois os datasets. Eu escolhi `DataScienceIdentity`

![challenge_panda](https://media.giphy.com/media/K9z3im98oo9Ve/giphy.gif)

In [36]:
free_responses = pd.read_csv('kaggle-survey-2017/freeformResponses.csv')

/usr/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (5,17,21,38,50) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [37]:
free_responses.head()

GenderFreeForm KaggleMotivationFreeForm CurrentJobTitleFreeForm  \
0            NaN                      NaN                     NaN   
1            NaN                      NaN                     NaN   
2            NaN                      NaN                 teacher   
3            NaN                      NaN                     NaN   
4            NaN                      NaN                     NaN   

  MLToolNextYearFreeForm MLMethodNextYearFreeForm  \
0                    NaN                      NaN   
1                    NaN                      NaN   
2                    NaN                      NaN   
3                    NaN                      NaN   
4                    NaN                      NaN   

  LanguageRecommendationFreeForm PublicDatasetsFreeForm  \
0                            NaN                    NaN   
1                            NaN                    NaN   
2                            NaN                    NaN   
3                            NaN                    NaN   
4                            NaN                    NaN   

            PersonalProjectsChallengeFreeForm  \
0                           Data manipulation   
1  I can't find time to practice consistently   
2                                         NaN   
3                    Connectivity/data fusion   
4                                         NaN   

  LearningPlatformCommunityFreeForm LearningPlatformFreeForm1  \
0                               NaN                       NaN   
1                               NaN                       NaN   
2                               NaN                   Meetups   
3                               NaN                       NaN   
4                               NaN                       NaN   

  LearningPlatformFreeForm2 LearningPlatformFreeForm3  \
0                       NaN                       NaN   
1                       NaN                       NaN   
2                       NaN                       NaN   
3                       NaN                       NaN   
4                       NaN                       NaN   

   LearningPlatformUsefulnessCommunitiesFreeForm  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  LearningPlatformUsefulnessFreeForm1Select  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   

   LearningPlatformUsefulnessFreeForm1SelectFreeForm  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

  LearningPlatformUsefulnessFreeForm2Select  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   

   LearningPlatformUsefulnessFreeForm2SelectFreeForm  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

  LearningPlatformUsefulnessFreeForm3Select  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   

   LearningPlatformUsefulnessFreeForm3SelectFree

In [38]:
free_responses.shape

(16716, 62)

In [39]:
free_responses.isnull().sum().sort_values()

WorkLibrariesFreeForm                                12221
InterestingProblemFreeForm                           12261
ImpactfulAlgorithmFreeForm                           12369
PersonalProjectsChallengeFreeForm                    13209
DataScienceIdentityFreeForm                          14299
PastJobTitlesFreeForm                                14622
CurrentJobTitleFreeForm                              15573
BlogsPodcastsNewslettersFreeForm                     15603
WorkMLTeamSeatFreeForm                               15788
EmployerIndustryOtherFreeForm                        15790
MajorFreeForm                                        15907
MLTechniquesFreeform                                 15963
KaggleMotivationFreeForm                             15970
MLSkillsFreeForm                                     16050
WorkToolsFreeForm1                                   16052
WorkDataTypeFreeForm                                 16055
EmployerSearchMethodOtherFreeForm                    160

In [40]:
free_index = free_responses['DataScienceIdentityFreeForm'].notnull()

In [41]:
identity_check = pd.DataFrame({'IdentityFree': free_responses[free_index]['DataScienceIdentityFreeForm'], 
              'IdentitySelect': multiple_choice[free_index]['DataScienceIdentitySelect']})

In [42]:
identity_check.shape

(2417, 2)

In [43]:
identity_check.isnull().sum()

IdentityFree        0
IdentitySelect    583
dtype: int64

Aparentemente não foi isso que aconteceu...

## Alterando o dataset original

In [44]:
df = multiple_choice.copy()

In [45]:
df['LearningDataScience'].value_counts()

Yes, I'm focused on learning mostly data science skills                  800
Yes, but data science is a small part of what I'm focused on learning    429
No, I am not focused on learning data science skills                      55
Name: LearningDataScience, dtype: int64

In [46]:
def replace_value(row):
    if row == "Yes, I'm focused on learning mostly data science skills":
        return "yes"
    elif row == "Yes, but data science is a small part of what I'm focused on learning":
        return "so so"
    elif row == "No, I am not focused on learning data science skills":
        return "no"

In [47]:
df['LearningDataScienceSimple'] = df['LearningDataScience'].apply(replace_value)

Agora podemos ver os novos valores desses campos

In [48]:
df['LearningDataScienceSimple'].value_counts()

yes      800
so so    429
no        55
Name: LearningDataScienceSimple, dtype: int64

In [49]:
df.shape

(16716, 229)

E agora que a outra coluna muito complexa não serve mais, podemos descartá-la

In [50]:
df.drop(['LearningDataScience'], axis=1, inplace=True)

In [51]:
df.shape

(16716, 228)

Pela quantidade de linhas no dataset percebemos que o `value_counts()` não retorna os valores nulos - e nós temos MUITOS valores nulos nessa coluna. Podemos utilizar um método do pandas para trocar os NAs por uma categoria nossa.

In [52]:
df['LearningDataScienceSimple'].fillna("did not answer the question", inplace=True)

In [53]:
df['LearningDataScienceSimple'].value_counts()

did not answer the question    15432
yes                              800
so so                            429
no                                55
Name: LearningDataScienceSimple, dtype: int64

Outra forma de alterar o dataset é utilizando funções _in place_ para isso utilizaremos o `lambda`

Por exemplo: E se eu quiser atualizar a idade dos participantes? O dataset foi coletado em 2017 e já estamos em 2018

In [54]:
df['NewAge'] = df['Age'].apply(lambda x: x + 1)

Vamos ver se funcionou? Vamos dar uma olhada nos primeiros 5 registros, com a coluna 'Age' e a 'NewAge' lado a lado

In [55]:
df[['Age','NewAge']][:5]

Age  NewAge
0   NaN     NaN
1  30.0    31.0
2  28.0    29.0
3  56.0    57.0
4  38.0    39.0

Ficou mais claro a proporção de pessoas que não responderam agora o/

### Desafio 3

Separar os datasets pelas pessoas que os responderam. Para isso você vai ter que carregar o dataset `schema.csv`.

Como uns datasets ficariam muito pequenos, sugiro que você utilize os seus conhecimentos recém adquiridos e crie 4 datasets distintos.

![panda_playground](https://media.giphy.com/media/ieaUdBJJC19uw/giphy.gif)

In [56]:
schema = pd.read_csv('kaggle-survey-2017/schema.csv')

In [57]:
schema.head()

Column                                           Question Asked
0      GenderSelect     Select your gender identity. - Selected Choice   All
1    GenderFreeForm  Select your gender identity. - A different ide...   All
2           Country          Select the country you currently live in.   All
3               Age                                   What's your age?   All
4  EmploymentStatus             What's your current employment status?   All

In [58]:
schema.Asked.value_counts()

CodingWorker       161
All                 70
Learners            41
Worker1              6
CodingWorker-NC      5
Worker               2
OnlineLearners       2
Non-worker           2
Non-switcher         1
Name: Asked, dtype: int64

In [59]:
coding_worker_column = schema[schema['Asked'] == 'CodingWorker']['Column']
all_column = schema[schema['Asked'] == 'All']['Column']
learners_column = schema[schema['Asked'] == 'Learners']['Column']
others_columns = schema[(~ schema['Asked'].isin(['CodingWorker', 'All', 'Learners']))]['Column']

In [60]:
all_multiple_selection_cols = [c for c in all_column.values if 'freeform' not in c.lower()]
coding_worker_multiple_selection_cols = [c for c in coding_worker_column.values if 'freeform' not in c.lower()]
learners_multiple_selection_cols = [c for c in learners_column.values if 'freeform' not in c.lower()]
others_multiple_selection_cols = [c for c in others_columns.values if 'freeform' not in c.lower()]

In [61]:
others_multiple_selection_cols

['StudentStatus',
 'LearningDataScience',
 'CodeWriter',
 'CareerSwitcher',
 'CurrentJobTitleSelect',
 'TitleFit',
 'CurrentEmployerType',
 'CoursePlatformSelect',
 'EmployerIndustry',
 'EmployerSize',
 'EmployerSizeChange',
 'EmployerMLTime',
 'EmployerSearchMethod']

In [63]:
all_multiple_selection = multiple_choice[all_multiple_selection_cols]
coding_worker_multiple_selection = multiple_choice[coding_worker_multiple_selection_cols]
learners_multiple_selection = multiple_choice[learners_multiple_selection_cols]
others_multiple_selection = multiple_choice[others_multiple_selection_cols]

In [64]:
print(all_multiple_selection.shape)
print(coding_worker_multiple_selection.shape)
print(learners_multiple_selection.shape)
print(others_multiple_selection.shape)

(16716, 43)
(16716, 137)
(16716, 35)
(16716, 13)


Lembrando que nós jogamos fora quem era 'free_form' - Logo, haverão sempre menos colunas do que vimos antes na quantidade por tipo de pessoa que respondeu.

In [68]:
all_multiple_selection.to_csv('all_multiple_selection.csv')
coding_worker_multiple_selection.to_csv('coding_worker_multiple_selection.csv')
learners_multiple_selection.to_csv('learners_multiple_selection.csv')
others_multiple_selection.to_csv('others_multiple_selection.csv')